### PR TITLE
Add AWS Batch runner

### PR DESCRIPTION
**Pull request by qiagu**
Adds the AWS batch runner to Galaxy.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/jobs/runners/aws.py [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception as e:
+            LOGGER.error(unicodify(e))
+            raise
+
+    return wrapper
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/aws.py]: You don't seem to be catching anything here, apart from duplicating the stringifying error message of the exception this doesn't do anything, so maybe remove it ? Or did you mean to ignore boto exceptions ?

File: lib/galaxy/jobs/runners/aws.py [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+try:
+    import boto3
+
+except ImportError as e:
+    boto3 = None
+    BOTO3_IMPORT_MSG.format(msg=unicodify(e))
+
+
+__all__ = ("AWSBatchJobRunner",)
+LOGGER = logging.getLogger(__name__)
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/aws.py]:
```suggestion
log = logging.getLogger(__name__)
```

File: lib/galaxy/jobs/runners/aws.py [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+        super().__init__(app, nworkers, **kwargs)
+        session = boto3.Session(
+            aws_access_key_id=self.runner_params.get("aws_access_key_id") or None,
+            aws_secret_access_key=self.runner_params.get("aws_secret_access_key") or None
+        )
+        self._batch_client = session.client("batch")
+
+    @handle_exception_call
+    def queue_job(self, job_wrapper):
+        LOGGER.debug(f"Starting queue_job for job {job_wrapper.get_id_tag()}")
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/aws.py]:
```suggestion
        log.debug(f"Starting queue_job for job {job_wrapper.get_id_tag()}")
```

File: lib/galaxy/jobs/runners/aws.py [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+        if not self.prepare_job(job_wrapper, include_metadata=False, modify_command_for_container=False):
+            LOGGER.debug(f"Not ready {job_wrapper.get_id_tag()}")
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/aws.py]:
```suggestion
            log.debug(f"Not ready {job_wrapper.get_id_tag()}")
```

File: lib/galaxy/jobs/runners/aws.py [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+            h.update(smart_str(k + str(v)))
+        queue_name = destination_params.get("job_queue").rsplit("/", 1)[-1]
+
+        jd_name = f"galaxy_tool__{tool_id}__{h.hexdigest()}__{queue_name}"
+        res = self._batch_client.describe_job_definitions(jobDefinitionName=jd_name, status="ACTIVE")
+        if not res["jobDefinitions"]:
+            jd_arn = self._register_job_definition(jd_name, container_image, destination_params)
+        else:
+            jd_arn = res["jobDefinitions"][0]["jobDefinitionArn"]
+            LOGGER.debug(f"Found existing job definition: {jd_name}.")
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/aws.py]:
```suggestion
            log.debug(f"Found existing job definition: {jd_name}.")
```

File: lib/galaxy/jobs/runners/aws.py [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+
+        return jd_arn
+
+    def _register_job_definition(self, jd_name, container_image, destination_params):
+        LOGGER.debug(f"Registering a new job definition: {jd_name}.")
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/aws.py]:
```suggestion
        log.debug(f"Registering a new job definition: {jd_name}.")
```

File: lib/galaxy/jobs/runners/aws.py [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+        )
+
+        assert res["ResponseMetadata"]["HTTPStatusCode"] == 200
+        return res["jobDefinitionArn"]
+
+    def _submit_job(self, job_def, job_wrapper, destination_params):
+        job_name = self.JOB_NAME_PREFIX + job_wrapper.get_id_tag()
+        command_script_path = self.write_command(job_wrapper)
+
+        LOGGER.info(f"Submitting job {job_name} to AWS Batch.")
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/aws.py]:
```suggestion
        log.info(f"Submitting job {job_name} to AWS Batch.")
```

**Review by mvdbeek — COMMENTED** [review]
Could you also write a bit of documentation on how to set it up and how to use it ?

File: lib/galaxy/jobs/runners/aws.py (hunk 2 above)
Comment (qiagu) [reply lib/galaxy/jobs/runners/aws.py]: Thanks @mvdbeek

File: doc/source/admin/cluster.md [hunk 8]
```diff
@@ -314,6 +315,63 @@ Torque attributes can be defined in either their short (e.g. [qsub(1B)](http://c
 
 Most options available to `qsub(1b)` and `pbs_submit(3b)` are supported.  Exceptions include `-o/Output_Path`, `-e/Error_Path`, and `-N/Job_Name` since these PBS job attributes are set by Galaxy.
 
+## AWS Batch
+
+Runs jobs via the [AWS Batch](https://aws.amazon.com/batch/). Built on top of AWS Elastic Container Service (ECS), AWS Batch enables users to run hundreds of thousands of jobs with simple configuration.
```
Comment (nuwang) [inline doc/source/admin/cluster.md]:
```suggestion
Runs jobs via [AWS Batch](https://aws.amazon.com/batch/). Built on top of AWS Elastic Container Service (ECS), AWS Batch enables users to run hundreds of thousands of jobs with little configuration.
```

File: doc/source/admin/cluster.md [hunk 9]
```diff
@@ -314,6 +315,63 @@ Torque attributes can be defined in either their short (e.g. [qsub(1B)](http://c
...
+
+#### Dependencies
+
+AWS Batch job runner requirs AWS Elastic File System (EFS) being mounted as a shared file system that enables Galaxy and job containers to read and write files. In the best pratice, Galaxy is installed an AWS EC2 instance and an EFS is mounted to the EC2 as a local drive. Job-related paths, such as objects, jobs_directory, tool_directory and so on, need to be placed on the EFS drive.
```
Comment (nuwang) [inline doc/source/admin/cluster.md]:
```suggestion
The AWS Batch job runner requires an AWS Elastic File System (EFS) to be mounted as a shared file system that enables Galaxy and job containers to read and write files. Galaxy is first installed on an AWS EC2 instance with EFS mounted as a local drive. Job-related paths, such as objects, jobs_directory, tool_directory and so on, need to be placed on the EFS drive.
```

File: doc/source/admin/cluster.md [hunk 10]
```diff
@@ -314,6 +315,63 @@ Torque attributes can be defined in either their short (e.g. [qsub(1B)](http://c
...
+In addition, Galaxy admin needs to configure Batch compute environment, Batch job queue and proper AWS IAM roles, and provision them as destination parameters.
+AWS Batch job runner requires [boto3](https://pypi.org/project/boto3/) installed in Galaxy environment.
+
+#### Parameters and Configuration
+
+AWS Batch job runner sends jobs to Batch compute environment that is composed of either Fargate or EC2. While Fargate provides a series of lightweigt compute resources (up to 4 vcpu and 30 GB memeory), the EC2 offers more abroad choices. With `auto_platform` enabled, this runner supports mapping to the best fit type of resources based on the provisioned `vcpu` and `memory`, i.e., Fargate is preferred over EC2 when `vcpu` and `memory` don't go beyond the limits (4 and 30 gb, respectively). If the power of `GPU` is needed for a destination, a job queue built on top of GPU-enabled compute environment must be provisoned.
```
Comment (nuwang) [inline doc/source/admin/cluster.md]:
```suggestion
In addition, a Galaxy admin needs to configure a Batch compute environment, a Batch job queue and proper AWS IAM roles, and specify them as destination parameters.
AWS Batch job runner requires [boto3](https://pypi.org/project/boto3/) to be installed in Galaxy's environment.

#### Parameters and Configuration

AWS Batch job runner sends jobs to Batch compute environment that is composed of either Fargate or EC2. While Fargate provides a series of lightweight compute resources (up to 4 vcpu and 30 GB memory), EC2 offers broader choices. With `auto_platform` enabled, this runner supports mapping to the best fit type of resources based on the requested `vcpu` and `memory`, i.e., Fargate is preferred over EC2 when `vcpu` and `memory` are below the limits (4 and 30 gb, respectively). If `GPU` computing is needed for a destination, a job queue built on top of a GPU-enabled compute environment must be provisioned.
```

File: lib/galaxy/config/sample/job_conf.xml.sample_advanced [hunk 11]
```diff
@@ -1112,7 +1123,41 @@
             <param id="max_retries">0</param>
             <!-- Number of retries to attempt if a command returns a non-zero status -->
         </destination>
-
+        <destination id="aws_batch_auto" runner="aws_batch">
+            <param id="docker_enabled">true</param>
+            <!-- `docker_enabled = true` is always required -->
+            <param id="job_queue">arn_for_Fargate_job_queue, arn_for_EC2_job_queue</param>
+            <!-- Fargete and non-GPU EC2 -->
```
Comment (nuwang) [inline lib/galaxy/config/sample/job_conf.xml.sample_advanced]:
```suggestion
            <!-- Fargate and non-GPU EC2 -->
```

File: doc/source/admin/cluster.md (hunk 9 above)
Comment (qiagu) [reply doc/source/admin/cluster.md]: @nuwang Thanks for your great help on my English. For `Galaxy is first installed on an AWS EC2 instance with EFS mounted as a local drive`, my understanding is that installation of Galaxy on EC2 is not necessary once EFS is able to be mounted. The AWS direct connect seems to allow mounting of EFS to happen outside of AWS, although no tests have been done yet.

Comment (nuwang) [reply doc/source/admin/cluster.md]: Sure, makes sense. Please reword if you think it's appropriate.

File: lib/galaxy/jobs/runners/aws.py [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+    be able to mount at multiple places with read and write capabilities, Galaxy and Batch
+    containers share the same EFS drive as a local device. Sample configurations can be found
+    in `config/job_conf.xml.sample_advanced`.
+    """
+
+    runner_name = "AWSBatchRunner"
+    RUNNER_PARAM_SPEC_KEY = "runner_param_specs"
+    JOB_NAME_PREFIX = "galaxy-"
+    # AWS Batch queries up to 100 jobs at once.
+    MAX_JOBS_PER_QUERY = 100
```
Comment (guerler) [inline lib/galaxy/jobs/runners/aws.py:96]: Thank you @qiagu for working on this. Can we add this parameter to the configuration options?

Comment (qiagu) [reply lib/galaxy/jobs/runners/aws.py:96]: The number/limit is managed by AWS. I don't think AWS changes the number often. Also, I imagine performance would be compromised if we change it to a smaller number. Is it really necessary to allow Galaxy admin to control it via job configuration?

**Review by guerler — APPROVED** [review]
Works well for me. Awesome work.

**Comment by dannon** [comment]
@qiagu Thanks for all your work here -- this is working really great in my testing on AWS.

**Review by dannon — APPROVED** [review]

**Pull request merged**